File: schemas/file-management-document-storage/drizzle/file_metadata_values.ts
// file_metadata_values: Custom metadata values per file — each row stores one field's value for one file.
// See README.md for full design rationale.

import {
  pgTable,
  uuid,
  text,
  timestamp,
  index,
  uniqueIndex,
} from "drizzle-orm/pg-core";
import { files } from "./files";
import { fileMetadataFields } from "./file_metadata_fields";

export const fileMetadataValues = pgTable(
  "file_metadata_values",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    fileId: uuid("file_id")
      .notNull()
      .references(() => files.id, { onDelete: "cascade" }),
    fieldId: uuid("field_id")
      .notNull()
      .references(() => fileMetadataFields.id, { onDelete: "cascade" }),

    // The actual value, stored as text regardless of field_type.
    // Application validates against field_type before saving.
    // Null means "explicitly empty".
    value: text("value"),

    createdAt: timestamp("created_at", { withTimezone: true })
      .notNull()
      .defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
      .notNull()
      .defaultNow()
      .$onUpdateFn(() => new Date()),
  },
  (table) => [
    // One value per field per file.
    uniqueIndex("uq_file_metadata_values_file_id_field_id").on(
      table.fileId,
      table.fieldId
    ),

    // "All files where invoice_number = 'INV-2024-001'."
    index("idx_file_metadata_values_field_id_value").on(
      table.fieldId,
      table.value
    ),
  ]
);

File: schemas/file-management-document-storage/spacetimedb/file_metadata_values.rs
// file_metadata_values: Custom metadata values per file — each row stores one field's value for one file.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = file_metadata_values, public)]
pub struct FileMetadataValue {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub file_id: String, // FK → files.id (cascade delete)

    #[index(btree)]
    pub field_id: String, // FK → file_metadata_fields.id (cascade delete)

    // The actual value, stored as text regardless of field_type.
    // Application validates against field_type before saving.
    // None means "explicitly empty".
    pub value: Option<String>,

    pub created_at: Timestamp,
    pub updated_at: Timestamp,

    // unique(file_id, field_id) — enforced at application level.
    // index(field_id, value) — composite index enforced at application level.
}
